Fixing overwrite and save slice permissions for a give role

sid88in opened the pull request:
@mistercrunch - I have fixed and tested slice save and slice overwrite permissions. Now a role can have flexibility to overwrite or save a slice. It will help in use cases where admin doesn't want users to overwrite a slice to mess up the dashboard.

Commit:
Fixing overwrite and save slice permissions for a give role

sid88in commented:
It fixes #143

Commit:
fix function name - build failed

mistercrunch commented on caravel/views.py:
```diff
@@ -468,7 +468,7 @@ def explore(self, datasource_type, datasource_id):
         action = request.args.get('action')
         if action in ('save', 'overwrite'):
-            return self.save(request.args, slc)
+            return self.saveoroverwriteslice(request.args, slc)
```
`underscore_between_words_please`

mistercrunch commented on caravel/views.py:
```diff
@@ -561,13 +557,28 @@ def save(self, args, slc):
+
+        return redirect(slc.slice_url)
+
+    @has_access
```
Oh I think this will create new view related permissions, where we should instead use the already defined model permissions using `if self.appbuilder.sm.has_access('can_add', 'Slices'):` and 
`if self.appbuilder.sm.has_access('can_edit', 'Slices'):`

Commit:
fix function name and test user permissions

sid88in commented:
1) Fixed function name.
2) Yes current changes will create new permissions. To avoid that, I have made new change. 

For admin self.appbuilder.sm.has_access('can_edit', 'Slices') and self.appbuilder.sm.has_access('can_add', 'Slices') permissions are False. The only permission available to admin on Slices on has menu access. I have changed the view name to SliceModelView instead of Slices and it works. Tested.

mistercrunch commented:
Nice, ideally we'd disable the buttons in the UI as well. All you have to do is pass the same `has_access` calls as `can_add` and `can_edit` to the template, and sometime like `{{ "disabled" if can_add }}` in the `class` of the button

Commit:
disable the button in the UI

sid88in commented:
Thats a good catch. I thought about the same but didn't find any implementation in the current code to disable UI features based on access controls. I have updated and tested the code for disabling the button (basically calling appbuilder functions in the template). I don't see a need for {{ "disabled" if can_add }}. We can simply add it in the if condition to show the button or not. And now the backend handles the same as well. Let me know if you need further modifications in the way its implemented in this PR. I will then create another PR to do the same for some other important features like - save dashboard, download csv.

sid88in commented:
Fyi - I tried replacing appbuilder.sm.has_access('can_edit', 'SliceModelView') with can_edit and it returned false. After reading template documentation I noticed one correct way to implement this is by calling objects passed in the render_template function. I don't see any current definition of can_edit and can_add in the current template code.

Commit:
fix build error - characters too long in 1 line

mistercrunch commented:
I meant you'd pass `can_add=self.appbuilder.sm.has_access('can_add', 'Slices')` to the template (same for `can_edit`), and use them to disable the buttons. I prefer disabling the buttons to hiding them, making it clear that the option exists and has not been granted.

sid88in commented:
in the latest changes, I did the same. Im passing these parameters in the template to disable the buttons. So I think we are on the same page. Having said that I can remove these conditions from views.py (which seem redundant). What do you say?

sid88in commented:
or may be you meant passing the values of can_edit and can_add in the template from explore function in the views.py. I think thats the same as replacing can_edit and can_add by self.appbuilder.sm.has_access('can_add', 'SliceModelView) in the template itself. Whatever you prefer :)

sid88in commented:
I think you overlooked my first comment ;)
Pasting from above:

Thats a good catch. I thought about the same but didn't find any implementation in the current code to disable UI features based on access controls. I have updated and tested the code for disabling the button (basically calling appbuilder functions in the template). I don't see a need for {{ "disabled" if can_add }}. We can simply add it in the if condition to show the button or not. And now the backend handles the same as well. Let me know if you need further modifications in the way its implemented in this PR. I will then create another PR to do the same for some other important features like - save dashboard, download csv.

sid88in commented:
<img width="338" alt="screen shot 2016-04-11 at 7 52 30 am" src="https://cloud.githubusercontent.com/assets/1587005/14431411/6546fa4e-ffba-11e5-93fd-c56f60df52d0.png">

sid88in commented:
<img width="356" alt="screen shot 2016-04-11 at 7 53 13 am" src="https://cloud.githubusercontent.com/assets/1587005/14431428/7841c25a-ffba-11e5-8f2d-9804f5be66d5.png">

mistercrunch commented:
From my previous comment: " I prefer disabling the buttons to hiding them"

Commit:
try to disable button on the UI

sid88in commented:
My bad. I didn't know the difference :/ I am testing the disabling feature on the UI - http://www.w3schools.com/bootstrap/tryit.asp?filename=trybs_button_active&stacked=h 

I

sid88in commented:
I tested disable feature. It show a stop sign and grey color if the access is not granted but user is still able to click on it (overwrite or save as). The operation is not successful and he can't overwrite or save a slice unless he has access. I am looking into disabling clicking on these buttons if there is no access. 

Testing disabling onlick on overwrite and save in explore.js

https://github.com/airbnb/caravel/blob/807d68631273a5a9f0e9c460cdedf9da19370153/caravel/assets/javascripts/explore.js#L279

mistercrunch commented:
All you need to do is to add "disabled" to the classes if the permission isn't there
```
<a class="btn btn-default {{ 'disabled' if not can_edit else '' }}">
```
Of course that means you need to pass `can_edit=appbuilder.sm.has_access('can_edit', 'SliceModelView')` to the render function.

Commit:
disable cursor in caravel css. You wont be able to click anymore if no access

Commit:
fix build issues

Commit:
Merge branch 'master' into slice_permissions

sid88in committed:
Merge branch 'master' into slice_permissions

Commit:
fix build errors! god bless me

Commit:
disable main features in dashboard and slice

Commit:
Merge branch 'master' of https://github.com/airbnb/caravel into slice_permissions

# Conflicts:
#	caravel/views.py

Commit:
fix build issues

Commit:
Merge branch 'master' into slice_permissions

mistercrunch commented:
Awesome! Mergin'
